src/js/components/FrameworkConfigurationForm.js

```diff
@@ -0,0 +1,337 @@
+      };
+    });
+  }
+
+  getUiSchema() {
+    const { formData, focusField, activeTab } = this.props;
+
+    const uiSchema = {
+      [activeTab]: {
+        [focusField]: { "ui:autofocus": true }
+      }
+    };
+
+    // hide all tabs not selected
+    Object.keys(formData).forEach(tabName => {
+      if (tabName !== activeTab) {
+        if (uiSchema[tabName] == null) {
+          uiSchema[tabName] = {};
+        }
+        uiSchema[tabName]["classNames"] = "hidden";
+      }
+    });
+
+    return uiSchema;
+  }
+
+  handleJSONChange(formData) {
+    this.props.onFormDataChange(formData);
+  }
+
+  writeErrors(formData, schema, isRequired, errors) {
+    if (Util.isObject(formData)) {
+      Object.keys(formData).forEach(property => {
+        this.writeErrors(
+          formData[property],
+          schema.properties[property],
+          schema.required && schema.required.includes(property),
+          errors[property]
+        );
+      });
+
+      return false;
+    }
+
+    if (isRequired && formData === "") {
+      let message = `Expecting a ${schema.type} here`;
+      if (schema.default) {
+        message = `Expecting a ${schema.type} here, eg: ${schema.default}`;
+      }
+      errors.addError(message);
+    }
+  }
+
+  getTotalErrorsForLevel(errorSchemaLevel) {
+    if (!errorSchemaLevel.__errors) {
+      return 0;
+    }
+    // only show one error per field in badge count
+    const currentLevelErrors = errorSchemaLevel.__errors.length > 0 ? 1 : 0;
+
+    return Object.keys(errorSchemaLevel)
+      .map(key => this.getTotalErrorsForLevel(errorSchemaLevel[key]))
+      .reduce((a, b) => a + b, currentLevelErrors);
+  }
+
+  handleFormChange(form) {
+    const { formData, errorSchema } = form;
+
+    const formErrors = {};
+    Object.keys(errorSchema).forEach(tab => {
+      formErrors[tab] = this.getTotalErrorsForLevel(errorSchema[tab]);
+    });
+
+    this.props.onFormErrorChange(formErrors);
+    this.props.onFormDataChange(formData);
+    this.setState({ errorSchema });
+  }
+
+  validate(formData, errors) {
+    const { packageDetails } = this.props;
+
+    this.writeErrors(formData, packageDetails.getConfig(), false, errors);
+
+    return errors;
+  }
+
+  formatErrorForJSON(errorSchema, path) {
+    if (!errorSchema.__errors) {
+      return [];
+    }
+    const currentLevelErrors = errorSchema.__errors.map(message => {
+      return { message, path };
+    });
+
+    return Object.keys(errorSchema)
+      .map(key => this.formatErrorForJSON(errorSchema[key], path.concat(key)))
+      .reduce((a, b) => a.concat(b), currentLevelErrors);
+  }
+
+  getErrorsForJSONEditor() {
+    const { errorSchema } = this.state;
+
+    if (!errorSchema) {
+      return [];
+    }
+
+    return this.formatErrorForJSON(errorSchema, []);
+  }
+
+  // transforms the errors from validating against the JSON schema
+  transformErrors(errorSchema) {
+    const errorSchemaCopy = [];
+
+    errorSchema.map(error => {
```

Comment:
☝️ Please adjust the transformation to either use `forEach` or even better to _properly_ use `map`. You're mutating `errorSchemaCopy` instead of mapping the values – I assume you did so to remove unwanted errors. You can either `filter` the list of errors before/after mapping or use reduce to do sone in one go. 

```js
  transformErrors(errorSchema) {
    // Filter "type" errors as they're caught by the validate() function and
    // capitalize error messages
    return errorSchema
      .filter(({ name }) => name !== "type")
      .map(({ message, ...rest }) => {
        return { message: StringUtil.capitalize(message), ...rest };
      });
  }
```

Reply:
Yes, much nicer. It was funny for me to read my very badly implemented version using `map`... 👍